tests/functional/test_install.py

```diff
@@ -1527,3 +1528,102 @@ def test_target_install_ignores_distutils_config_install_prefix(script):
         (target - script.base_path) in result.files_created and
         (prefix - script.base_path) not in result.files_created
     ), str(result)
+
+
+@pytest.mark.skipif("sys.platform != 'win32'")
+@pytest.mark.parametrize('pip_name', [
+    'pip',
+    'pip{}'.format(sys.version_info[0]),
+    'pip{}.{}'.format(*sys.version_info[:2]),
+    'pip.exe',
+    'pip{}.exe'.format(sys.version_info[0]),
+    'pip{}.{}.exe'.format(*sys.version_info[:2])
+])
+def test_protect_pip_from_modification_on_windows(script, pip_name):
+    """
+    Test ``pip install --upgrade pip`` is raised an error on Windows.
+    """
+    command = [pip_name, 'install', '--upgrade', 'pip']
+    result = script.run(*command, expect_error=True)
+    assert result.returncode != 0
+    new_command = [sys.executable, '-m', 'pip'] + command[1:]
```

Comment:
I've changed the assertion and checked the result.
```
>       assert result.returncode == 0, str(result)
E       AssertionError: Script result: pip install --upgrade pip
E           return code: 1
E         -- stderr: --------------------
E         ERROR: To modify pip, please run the following command:
E         /Users/albert/Projects/pip/.tox/py37/bin/python -m pip install --upgrade pip
E
E         -- stdout: --------------------
E         Requirement already up-to-date: pip in /Users/albert/Projects/pip/.tox/py37/lib/python3.7/site-packages (19.3.dev0)
```

Indeed, it raises an error "To modify pip, please run the following command" even if pip "already up-to-date".

Would you like to open an issue for that case?

Reply:
> This might make a plain pip invocation unusable on Windows depending on how common it is to have it listed as a dependency, since any pip appearing in the dependency graph would result in an error message IIUC.

It's pretty rare but I'm okay with starting like this and later, moving over to being more "strict" about when the message is printed based on user feedback and developer availability.

A tracking issue would be helpful.

Reply:
Okay, i'll raise an issue for that case after this PR being merged.

Reply:
Done #6924.